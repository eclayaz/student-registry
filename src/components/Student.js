import React, { useState, useEffect } from "react";
import EnrollStudent from "./EnrollStudent";

function Student({ student, subjects, enrollStudentHandler }) {
  const getSubjectsWithState = () => {
    return subjects.map((subject) => {
      return {
        name: subject.name,
        value: student.subjects.includes(subject.name) ? true : false,
      };
    });
  };

  const [subjectState, setSubjectState] = useState(getSubjectsWithState());

  const [showEnroll, setShowEnroll] = useState(false);
  useEffect(() => {
    return function cleanup() {
      setSubjectState(getSubjectsWithState());
    };
  }, [showEnroll]);

  const handleEnrollClose = () => {
    setShowEnroll(false);
  };

  const handleEnrollShow = () => {
    setShowEnroll(true);
  };

  const handleEnrollSubmit = async () => {
    let subjectsToEnroll = subjectState.reduce((filtered, subject) => {
      if (subject.value) {
        filtered.push(subject.name);
      }
      return filtered;
    }, []);

    try {
      await enrollStudentHandler(student._id, subjectsToEnroll);
      setShowEnroll(false);
    } catch (err) {
      console.log(err.message);
    }
  };

  const subjectCheckHandler = (event) => {
    setSubjectState(
      subjectState.map((subject) => {
        if (subject.name === event.target.name) {
          subject.value = event.target.checked;
        }
        return subject;
      })
    );
  };

  return (
    <tr key={student._id}>
      <td data-testid="name">{student.name}</td>
      <td data-testid="gender">{student.gender}</td>
      <td data-testid="address">{student.address}</td>
      <td data-testid="contactNumber">{student.contactNumber}</td>
      <td>
        <ul data-testid="subjects">
          {student.subjects.map((subject) => {
            return <li key={subject}>{subject}</li>;
          })}
        </ul>
      </td>
      <td>
        <EnrollStudent
          subjects={subjectState}
          handleEnrollSubmit={handleEnrollSubmit}
          subjectCheckHandler={subjectCheckHandler}
          handleEnrollClose={handleEnrollClose}
          handleEnrollShow={handleEnrollShow}
          showEnroll={showEnroll}
        />
      </td>
    </tr>
  );
}

export default Student;
