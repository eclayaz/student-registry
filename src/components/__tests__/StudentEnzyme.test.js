import React from "react";
import { shallow } from "enzyme";

import Student from "../Student";

const subjects = [
  {
    _id: "5ea1e6072113ead1c4cc8d82",
    name: "Maths",
  },
  {
    _id: "5ea1e61a2113ead1c4cc8d83",
    name: "Science",
  },
];

const student = {
  _id: "5ea1dfc56d485cd446b54d90",
  name: "Nadeesha Dilruwna",
  gender: "Male",
  address: "Panadura",
  contactNumber: "766838371",
  subjects: ["Science"],
};

it("renders with student data", () => {
  let wrapped = shallow(<Student subjects={subjects} student={student} />);
  expect(wrapped).toMatchSnapshot();
});
