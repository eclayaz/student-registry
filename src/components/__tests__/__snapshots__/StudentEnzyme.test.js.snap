// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders with student data 1`] = `
<tr
  key="5ea1dfc56d485cd446b54d90"
>
  <td
    data-testid="name"
  >
    Nadeesha Dilruwna
  </td>
  <td
    data-testid="gender"
  >
    Male
  </td>
  <td
    data-testid="address"
  >
    Panadura
  </td>
  <td
    data-testid="contactNumber"
  >
    766838371
  </td>
  <td>
    <ul
      data-testid="subjects"
    >
      <li
        key="Science"
      >
        Science
      </li>
    </ul>
  </td>
  <td>
    <EnrollStudentPopup
      handleEnrollClose={[Function]}
      handleEnrollShow={[Function]}
      handleEnrollSubmit={[Function]}
      showEnroll={false}
      subjectCheckHandler={[Function]}
      subjects={
        Array [
          Object {
            "name": "Maths",
            "value": false,
          },
          Object {
            "name": "Science",
            "value": true,
          },
        ]
      }
    />
  </td>
</tr>
`;
